File: src/main/java/io/zeitwert/jooq/persistence/AggregatePersistenceProviderBase.java
package io.zeitwert.jooq.persistence;

import java.util.HashMap;
import java.util.Map;

import org.jooq.DSLContext;

import io.zeitwert.ddd.aggregate.model.Aggregate;
import io.zeitwert.ddd.aggregate.model.AggregatePersistenceProvider;
import io.zeitwert.ddd.aggregate.model.AggregateRepository;
import io.zeitwert.ddd.app.service.api.AppContext;

public abstract class AggregatePersistenceProviderBase<A extends Aggregate> implements AggregatePersistenceProvider<A> {

	private final Class<? extends Aggregate> intfClass;
	private final DSLContext dslContext;
	private final Map<String, Object> dbConfigMap = new HashMap<>();

	public AggregatePersistenceProviderBase(Class<? extends Aggregate> intfClass, DSLContext dslContext) {
		this.intfClass = intfClass;
		this.dslContext = dslContext;
	}

	@Override
	public final Class<? extends Aggregate> getEntityClass() {
		return this.intfClass;
	}

	public final DSLContext dslContext() {
		return this.dslContext;
	}

	public final Map<String, Object> dbConfigMap() {
		return this.dbConfigMap;
	}

	@SuppressWarnings("unchecked")
	public final AggregateRepository<A, ?> getRepository() {
		return (AggregateRepository<A, ?>) AppContext.getInstance().getRepository(this.intfClass);
	}

}

File: fm-dddrive/src/main/java/io/dddrive/core/ddd/model/base/AggregatePersistenceProviderBase.kt
package io.dddrive.core.ddd.model.base

import io.dddrive.core.ddd.model.Aggregate
import io.dddrive.core.ddd.model.AggregatePersistenceProvider
import io.dddrive.core.ddd.model.RepositoryDirectory
import io.dddrive.core.ddd.model.RepositoryDirectorySPI

abstract class AggregatePersistenceProviderBase<A : Aggregate>(
	intfClass: Class<A>,
) : AggregatePersistenceProvider<A> {

	val directory: RepositoryDirectory get() = RepositoryDirectory.getInstance()

	init {
		(directory as RepositoryDirectorySPI).addPersistenceProvider(intfClass, this)
	}

}
